### PR TITLE
Add some simple unit tests for SettingsHandler

### PR DESCRIPTION
**Pull request by nlebeck**
This is my first attempt at adding unit tests in the Dolphin codebase. It looks like most of the unit test files don't have many comments, but please let me know if test plan comments would be helpful here, or if you have any feedback on the scope or style of the tests.

### CODE REVIEW

**Comment by Dentomologist** [comment]
@dolphin-emu-bot rebuild

**Comment by JosJuice** [comment]
Instead of having tests that both encrypt and decrypt and then check the result, I would suggest having tests that do just encryption and just decryption. (Possibly in addition to ones that do both. `GetValueOnSameInstance` can be kept the way it is, for instance.) SettingsHandler is implementing a format defined by Nintendo, and we have to make sure we read and write files in a way that's compatible with Nintendo's files.

Additionally, if you're able to, it would be nice to have tests for the edge case implemented in PR #8704.

**Comment by nlebeck** [comment]
Sounds good, I'll make that change!

Adding test coverage for the edge case in PR #8704 seems fun. I'll either do that in here or leave it for a follow-up PR depending on how tricky it is.

**Comment by nlebeck** [comment]
Changed (most of) the tests to test encryption and decryption separately. PTAL! I'm not used to writing tests where the input/output data is raw u8 arrays, so let me know if there are any style concerns.

I left a TODO for PR #8704. I'd like to add that coverage, but I have pretty limited free time these days, so it might be a while before I get to it.

File: Source/UnitTests/Common/SettingsHandlerTest.cpp [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+  u8 expected_buffer[BUFFER_SIZE];
+  std::fill(expected_buffer, expected_buffer + BUFFER_SIZE, 0);
+  std::copy(BUFFER_CONTENTS_B.begin(), BUFFER_CONTENTS_B.end(), expected_buffer);
+
+  ASSERT_EQ(buffer.size(), BUFFER_SIZE);
+  for (size_t i = 0; i < buffer.size(); ++i)
+  {
+    EXPECT_EQ(buffer[i], expected_buffer[i]);
+  }
```
Comment (noahpistilli) [inline Source/UnitTests/Common/SettingsHandlerTest.cpp:68]: Same thing, just with `BUFFER_CONTENTS_B`.

File: Source/UnitTests/Common/SettingsHandlerTest.cpp [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+#include "Common/SettingsHandler.h"
+
+namespace
+{
+constexpr size_t BUFFER_SIZE = Common::SettingsHandler::SETTINGS_SIZE;
+
+// The encrypted bytes corresponding to the following settings, in order:
+//   "key" = "val"
+std::array<char, 9> BUFFER_CONTENTS_A = {'\x91', '\x91', '\x90', '\xEE', '\xD1',
```
[the remark this answers is not in the document]
Comment (nlebeck) [reply Source/UnitTests/Common/SettingsHandlerTest.cpp]: Done. (It's unfortunate that conversion of a character literal to `u8` is a narrowing conversion. Is there anything cleaner we can do here?)

File: Source/UnitTests/Common/SettingsHandlerTest.cpp [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+                                         '/',    '\xF0', '4',    'y'};
+
+// The encrypted bytes corresponding to the following settings, in order:
+//   "key1" = "val1"
+//   "key2" = "val2"
+//   "foo" = "bar"
+std::array<char, 31> BUFFER_CONTENTS_B = {
```
[the remark this answers is not in the document]
Comment (nlebeck) [reply Source/UnitTests/Common/SettingsHandlerTest.cpp]: Done.

File: Source/UnitTests/Common/SettingsHandlerTest.cpp (hunk 1 above)
Comment (nlebeck) [reply Source/UnitTests/Common/SettingsHandlerTest.cpp]: Done.

**Comment by nlebeck** [comment]
Thanks for the review!

File: Source/UnitTests/Common/SettingsHandlerTest.cpp [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+{
+  Common::SettingsHandler handler;
+  handler.AddSetting("key", "val");
+  Common::SettingsHandler::Buffer buffer = handler.GetBytes();
+
+  ASSERT_EQ(buffer.size(), BUFFER_SIZE);
+  for (size_t i = 0; i < BUFFER_CONTENTS_A.size(); ++i)
+  {
+    EXPECT_EQ(buffer[i], BUFFER_CONTENTS_A[i]);
+  }
```
Comment (JosJuice) [inline Source/UnitTests/Common/SettingsHandlerTest.cpp:38]: This loop could be replaced with calling `std::equal` or `std::memcpy`. Same for the other tests.

File: Source/UnitTests/Common/SettingsHandlerTest.cpp [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+TEST(SettingsHandlerTest, DecryptSingleSetting)
+{
+  Common::SettingsHandler::Buffer buffer;
+  std::copy(BUFFER_CONTENTS_A.begin(), BUFFER_CONTENTS_A.end(), buffer.begin());
+  Common::SettingsHandler handler(std::move(buffer));
```
Comment (JosJuice) [inline Source/UnitTests/Common/SettingsHandlerTest.cpp:45]: It's unfortunate that this little "dance" is needed. I would suggest adding an additional constructor to `SettingsHandler` so it accepts `const Buffer&` in addition to `Buffer&&`.

File: Source/UnitTests/UnitTests.vcxproj [hunk 6]
```diff
@@ -54,6 +54,7 @@
     <ClCompile Include="Common\FloatUtilsTest.cpp" />
     <ClCompile Include="Common\MathUtilTest.cpp" />
     <ClCompile Include="Common\NandPathsTest.cpp" />
+    <ClCompile Include="Common\SettingsHandlerTest.cpp" />
```
Comment (JosJuice) [inline Source/UnitTests/UnitTests.vcxproj:57]: The test must also be added to CMakeLists.txt, otherwise it won't run on platforms other than Windows.

**Review by JosJuice — COMMENTED** [review]
Please squash your commits into one.

File: Source/UnitTests/UnitTests.vcxproj (hunk 6 above)
Comment (nlebeck) [reply Source/UnitTests/UnitTests.vcxproj:57]: Done.

File: Source/UnitTests/Common/SettingsHandlerTest.cpp (hunk 4 above)
Comment (nlebeck) [reply Source/UnitTests/Common/SettingsHandlerTest.cpp]: Done (here and below).

File: Source/UnitTests/Common/SettingsHandlerTest.cpp (hunk 5 above)
Comment (nlebeck) [reply Source/UnitTests/Common/SettingsHandlerTest.cpp:42]: Two thoughts here:

1) The little "dance" is necessary not just because of the move constructor, but because a `Buffer` always has length 256. If we take a const ref, we'd still have to `std::copy` from our `BUFFER_CONTENTS` array into the passed-in `Buffer` somewhere, or add a bunch of padding to our hard-coded arrays.

2) Would the additional constructor be useful in the prod code? If not, and if we always want to avoid a copy when constructing a `SettingsHandler` in the prod code, my inclination would be to avoid adding it just for the sake of making the tests a bit more concise. I'm inspired here by the "DAMP, not DRY" philosophy in https://abseil.io/resources/swe-book/html/ch12.html, and a culture in my day job of avoiding adding test-only API surface when possible. (I don't want to apply style guidance I learned at my day job if the Dolphin project has a different philosophy, though.)

WDYT? I'll defer to your judgment!

**Comment by nlebeck** [comment]
Thanks for the review! Commits squashed.

File: Source/UnitTests/Common/SettingsHandlerTest.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (noahpistilli) [reply Source/UnitTests/Common/SettingsHandlerTest.cpp]: Rather than `'\x91'`, make it `0x91`. So hexadecimal integer rather than its char interpretation.

For the values that you can't easily convert by inspection ('B', '/', etc...), you can use a hex editor or Google to find its hexadecimal value.

[the remark this answers is not in the document]
Comment (nlebeck) [reply Source/UnitTests/Common/SettingsHandlerTest.cpp]: Ah right! Thanks.

File: Source/UnitTests/Common/SettingsHandlerTest.cpp (hunk 5 above)
Comment (nlebeck) [reply Source/UnitTests/Common/SettingsHandlerTest.cpp:39]: Actually, now that I think about more, I don't think moving a `std::array` of primitives actually avoids any copies (https://stackoverflow.com/a/28988913). So maybe this class should only have a constructor that takes `const Buffer&`, if a `Buffer` will always just be a type alias for an array.

I'll do whatever you think is appropriate in this PR, and I'm happy to put together a follow-up PR if we want to remove the rvalue refs in favor of const refs.

Comment (nlebeck) [reply Source/UnitTests/Common/SettingsHandlerTest.cpp]: Sigh, I read the `std::array` spec more closely and realized my point 1 above isn't true either. We can have the hardcoded arrays be length 256 and use list initialization to explicitly initialize their first few elements and value-initialize the rest. And then we can simply use the assignment operator instead of `std::copy` in the test bodies.

In the latest commit, the "dance" to create a temp buffer has been reduced to `Buffer buffer = BUFFER_A`. Seems sufficiently clean that there's no need to add a new constructor, but I'm happy to do it if you want.

Sorry about all the mistakes, and thanks for bearing with me!

Comment (JosJuice) [reply Source/UnitTests/Common/SettingsHandlerTest.cpp]: Sorry, I missed these comments. Let's go with what you have in the current version.

**Review by JosJuice — APPROVED** [review]

**Review by noahpistilli — APPROVED** [review]
